### PR TITLE
[3144] Fixed autocomplete namestartswithquery issue

### PR DESCRIPTION
**Pull request by defong**
### Context
Autocomplete namestartswithquery issue

### Changes proposed in this pull request
Fixed autocomplete namestartswithquery issue

### Guidance to review

This capture its failure state

https://github.com/DFE-Digital/register-trainee-teachers/pull/1717/commits/2e060071c34de03c8c54473bfe4f7cc598671f5f

Largely a refactoring to make it a bit more palatable
Additional tests was added to assert verification
https://github.com/DFE-Digital/register-trainee-teachers/pull/1717/commits/9494849b36e5c23e1bd2033b6b85ac7079569a72

### CODE REVIEW

**Comment by defong** [comment]
[Autocomplete - `nameStartsWithQuery` bug](https://trello.com/c/BIrkZGB9/3144-autocomplete-namestartswithquery-bug)

**Review by AbigailMcP — APPROVED** [review]

File: app/components/form_components/autocomplete/sort/calculateWeight.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+const exactMatch = (word, query) => (word === query)
+
+const startsWithReqExp = (query) => new RegExp('\\b' + query, 'i')
```
[the remark this answers is not in the document]
Comment (defong) [reply app/components/form_components/autocomplete/sort/calculateWeight.js]: done

File: app/components/form_components/autocomplete/sort/calculateWeight.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+// Returns an integer ranging from 0 (no match) to 100 (exact match).
+const calculateWeight = ({ name, synonyms }, query) => {
+  if (exactMatch(name, query)) return 100
+  if (synonymsExactMatch(synonyms, query)) return 75
+  if (startsWith(name, query)) return 60
+  if (synonymsStartsWith(synonyms, query)) return 50
+
+  const startsWithReqExps = query.split(/\s+/).map(startsWithReqExp)
+
+  if (wordsStartsWithQuery(name, startsWithReqExps)) return 25
```
[the remark this answers is not in the document]
Comment (defong) [reply app/components/form_components/autocomplete/sort/calculateWeight.js]: done

File: app/components/form_components/autocomplete/sort/calculateWeight.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+// NOTE: Determines how the query matches either an option's name/synonyms.
+// Returns an integer ranging from 0 (no match) to 100 (exact match).
+const calculateWeight = ({ name, synonyms }, query) => {
+  if (exactMatch(name, query)) return 100
+  if (synonymsExactMatch(synonyms, query)) return 75
+  if (startsWith(name, query)) return 60
+  if (synonymsStartsWith(synonyms, query)) return 50
+
+  const startsWithReqExps = query.split(/\s+/).map(startsWithReqExp)
```
[the remark this answers is not in the document]
Comment (defong) [reply app/components/form_components/autocomplete/sort/calculateWeight.js]: done

**Review by edwardhorsford — APPROVED** [review]
Looks good to me in Chrome. Think we might want to test this with some older browsers though.

**Pull request merged**